app/grandchallenge/reader_studies/views.py

```diff
@@ -1429,28 +1429,33 @@ def form_valid(self, form):
         return HttpResponseRedirect(self.get_success_url())
 
     def create_civ_for_value(
-        self, instance, ci_slug, new_value, assigned_civs
+        self, instance, ci, current_civ, new_value, assigned_civs
     ):
-        ci = ComponentInterface.objects.get(slug=ci_slug)
-        current_value = instance.values.filter(interface=ci).first()
-        if ci.is_json_kind and not ci.requires_file:
-            if current_value:
-                assigned_civs.append(current_value)
-            val = ComponentInterfaceValue.objects.create(
+        current_value = current_civ.value if current_civ else None
+        if new_value and current_value != new_value:
+            assigned_civs.append(current_civ)
+            civ = ComponentInterfaceValue.objects.create(
                 interface=ci, value=new_value
             )
-            instance.values.add(val)
-        elif isinstance(new_value, Image):
-            if not current_value or (
-                current_value and current_value.image != new_value
-            ):
-                assigned_civs.append(current_value)
-                civ, created = ComponentInterfaceValue.objects.get_or_create(
-                    interface=ci, image=new_value
-                )
-                if created:
-                    civ.full_clean()
-                instance.values.add(civ)
+            civ.full_clean()
+            instance.values.add(civ)
+        elif not new_value:
+            # if the new value is None, remove the old CIV from the display set
+            assigned_civs.append(current_civ)
```

Comment:
Sorry, trying to understand this code still. I remember being lots of issues with it which is why there are a lot of existing comments. Why this construction where values are added one by one, but collected and then removed in bulk by passing round a list? Surely it's better just to remove them directly with `instance.values.remove(current_civ)` - then you could remove all instances of `assigned_civs`, right? Or am I missing something?

Reply:
I also don't see why `instance` needs to be returned as it is not being modified, right?

Reply:
You're right. Removing them directly works and is definitely better.

Reply:
And passing the instance around also shouldn't be necessary. I'll double check, of course!